File: Retrofit/src/main/java/kim/uno/sample/retrofit/ObjectCallBack.java
package kim.uno.sample.retrofit;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class ObjectCallBack<T> implements Callback<T> {

    @Override
    final public void onResponse(Call<T> call, Response<T> response) {
        onResponse(call, response, null);
    }

    @Override
    final public void onFailure(Call<T> call, Throwable t) {
        onResponse(call, null, t);
    }
    
    public void onResponse(Call<T> call, Response<T> response, Throwable t) {
        
    }
}

File: Retrofit/src/main/java/kim/uno/sample/retrofit/ObjectCallBack.kt
package kim.uno.sample.retrofit

import kim.uno.kotlin.base.util.GsonUtil
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

interface ObjectCallBack<T> : Callback<String> {

    fun parseTo() : Class<T>

    override fun onResponse(call: Call<String>?, response: Response<String>?) {
        onResponse(GsonUtil.fromJson(response?.body()?: "", parseTo()), null)
    }

    override fun onFailure(call: Call<String>?, t: Throwable?) {
        onResponse(null, t)
    }

    fun onResponse(response: T?, t: Throwable?)

}
